File: frontend/src/api/mock/course.js
const mockData = [
  {
    id: 1,
    title: "Тематические",
    type: "section",
    is_root: true,
    parent_id: null,
    owner_id: 1,
    owner_username: "owner",
    children: [
      {
        id: 11,
        title: "Нейросети",
        type: "section",
        is_root: false,
        parent_id: 1,
        owner_id: 1,
        owner_username: "owner",
        children: [
          { id: 111, title: "Введение в нейросети", type: "course" },
          { id: 112, title: "CNN на практике", type: "course" },
          { id: 113, title: "RNN и последовательности", type: "course" },
        ],
      },
      {
        id: 12,
        title: "Классический ML",
        type: "section",
        is_root: false,
        parent_id: 1,
        owner_id: 1,
        owner_username: "owner",
        children: [
          {
            id: 121,
            title: "Линейная регрессия",
            type: "section",
            is_root: false,
            parent_id: 12,
            owner_id: 1,
            owner_username: "owner",
            children: [
              { id: 1211, title: "Основы линейных моделей", type: "course" },
              { id: 1212, title: "Градиентный спуск в регрессии", type: "course" },
            ],
          },
          { id: 122, title: "KNN", type: "course" },
          { id: 123, title: "Логистическая регрессия", type: "course" },
        ],
      },
    ],
  },
  {
    id: 2,
    title: "Олимпиады",
    type: "section",
    is_root: true,
    parent_id: null,
    owner_id: 1,
    owner_username: "owner",
    children: [
      {
        id: 21,
        title: "ВСОШ",
        type: "section",
        is_root: false,
        parent_id: 2,
        owner_id: 1,
        owner_username: "owner",
        children: [
          { id: 211, title: "Базовый трек", type: "course" },
          { id: 212, title: "Финальный тур", type: "course" },
        ],
      },
      {
        id: 22,
        title: "НТО ИИ",
        type: "section",
        is_root: false,
        parent_id: 2,
        owner_id: 1,
        owner_username: "owner",
        children: [
          { id: 221, title: "Подготовка к отбору", type: "course" },
          { id: 222, title: "Решение треков", type: "course" },
        ],
      },
      {
        id: 23,
        title: "Большие данные: ИИ",
        type: "section",
        is_root: false,
        parent_id: 2,
        owner_id: 1,
        owner_username: "owner",
        children: [
          { id: 231, title: "Data pipeline", type: "course" },
          { id: 232, title: "ML на больших данных", type: "course" },
        ],
      },
    ],
  },
  {
    id: 3,
    title: "Авторские",
    type: "section",
    is_root: true,
    parent_id: null,
    owner_id: 1,
    owner_username: "owner",
    children: [
      {
        id: 31,
        title: "В.М. Гуровиц",
        type: "section",
        is_root: false,
        parent_id: 3,
        owner_id: 1,
        owner_username: "owner",
        children: [
          { id: 311, title: "Анализ данных от А до Я", type: "course" },
          { id: 312, title: "Практикум по моделям", type: "course" },
        ],
      },
      {
        id: 32,
        title: "Вадим",
        type: "section",
        is_root: false,
        parent_id: 3,
        owner_id: 1,
        owner_username: "owner",
        children: [
          { id: 321, title: "ML интенсив", type: "course" },
          { id: 322, title: "Компьютерное зрение", type: "course" },
        ],
      },
    ],
  },
];

export function getCourses() {
  return new Promise(resolve => setTimeout(() => resolve(mockData), 200));
}

const flattenCourses = (nodes = []) => {
  const result = []
  nodes.forEach(node => {
    result.push(node)
    if (Array.isArray(node.children)) {
      result.push(...flattenCourses(node.children))
    }
  })
  return result
}

export function getCourse(courseId) {
  const numericId = Number(courseId)
  if (!Number.isFinite(numericId)) {
    return Promise.resolve(null)
  }
  const found = flattenCourses(mockData).find(item => item.id === numericId)
  if (!found) {
    return Promise.resolve(null)
  }
  return Promise.resolve({
    id: found.id,
    title: found.title,
    description: found.description || '',
    is_open: true,
    owner_id: 1,
    owner_username: 'owner',
    can_create_contest: true,
    can_manage_course: true,
    participants: [
      { id: 1, username: 'owner', role: 'teacher', is_owner: true },
      { id: 2, username: 'teacher2', role: 'teacher', is_owner: false },
      { id: 3, username: 'student', role: 'student', is_owner: false },
    ],
  })
}

export function createCourse(payload) {
  const id = Date.now()
  return Promise.resolve({ id, ...payload })
}

export function createSection(payload) {
  const id = Date.now()
  return Promise.resolve({ id, ...payload })
}

export function updateCourse(courseId, data) {
  return Promise.resolve({ id: Number(courseId), ...data })
}

export function deleteCourse(courseId) {
  return Promise.resolve({ success: true, deleted_id: Number(courseId) })
}

export function updateCourseParticipants(courseId, payload) {
  return Promise.resolve({ course_id: Number(courseId), ...payload, created: [], updated: [] })
}

export function removeCourseParticipants(courseId, usernames) {
  return Promise.resolve({ course_id: Number(courseId), removed: usernames || [] })
}

export function reorderCourseContests(courseId, contestIds) {
  return Promise.resolve({ course_id: Number(courseId), contest_ids: contestIds || [] })
}

File: backend/runner/views/course.py
from django.db.models import Count, Max
from django.http import JsonResponse
from django.shortcuts import get_object_or_404

from django.contrib.auth import get_user_model
from django.contrib.auth.decorators import login_required

from ..models import Contest, Course, CourseParticipant
from ..services.user_access import is_platform_admin

User = get_user_model()


def _contest_timing_payload(contest: Contest) -> dict:
    end_time = contest.get_end_time()
    return {
        "start_time": contest.start_time.isoformat() if contest.start_time else None,
        "end_time": end_time.isoformat() if end_time else None,
        "duration_minutes": contest.duration_minutes,
        "has_time_limit": bool(contest.start_time and contest.duration_minutes),
        "allow_upsolving": bool(contest.allow_upsolving),
        "time_state": contest.time_state(),
    }


def _course_is_teacher(course: Course, user) -> bool:
    if not user.is_authenticated:
        return False
    if is_platform_admin(user):
        return True
    if user.is_staff or user.is_superuser:
        return True
    if course.owner_id == user.id:
        return True
    return course.participants.filter(
        user=user, role=CourseParticipant.Role.TEACHER
    ).exists()


def _course_is_member(course: Course, user) -> bool:
    if not user.is_authenticated:
        return False
    if is_platform_admin(user):
        return True
    if user.is_staff or user.is_superuser:
        return True
    if course.owner_id == user.id:
        return True
    return course.participants.filter(user=user).exists()


def course_detail(request, course_id):
    if request.method != "GET":
        return JsonResponse({"detail": "Method not allowed"}, status=405)
    if not request.user.is_authenticated:
        return JsonResponse({"detail": "Authentication required"}, status=401)

    course = get_object_or_404(
        Course.objects.select_related("section", "owner").prefetch_related("participants__user"),
        pk=course_id,
    )
    is_admin = bool(is_platform_admin(request.user) or request.user.is_staff or request.user.is_superuser)
    is_owner = course.owner_id == request.user.id
    is_teacher = _course_is_teacher(course, request.user)
    is_member = _course_is_member(course, request.user)
    if not (is_admin or is_member or course.is_open):
        return JsonResponse({"detail": "Forbidden"}, status=403)

    participants = []
    if is_admin or is_member:
        participants = [
            {
                "id": participant.user_id,
                "username": participant.user.username,
                "role": participant.role,
                "is_owner": participant.is_owner,
            }
            for participant in course.participants.all()
            if not is_platform_admin(participant.user)
        ]

    return JsonResponse(
        {
            "id": course.id,
            "title": course.title,
            "description": course.description,
            "is_open": course.is_open,
            "section": course.section_id,
            "section_title": course.section.title,
            "owner_id": course.owner_id,
            "owner_username": course.owner.username if course.owner_id else None,
            "can_create_contest": bool(is_teacher),
            "can_manage_course": bool(is_owner or is_admin),
            "participants": participants,
        },
        status=200,
    )


def course_contests(request, course_id):
    if request.method != "GET":
        return JsonResponse({"detail": "Method not allowed"}, status=405)
    if not request.user.is_authenticated:
        return JsonResponse({"detail": "Authentication required"}, status=401)

    course = get_object_or_404(
        Course.objects.select_related("section", "owner").prefetch_related("participants__user"),
        pk=course_id,
    )
    is_admin = bool(is_platform_admin(request.user) or request.user.is_staff or request.user.is_superuser)
    is_member = _course_is_member(course, request.user)
    if not (is_admin or is_member or course.is_open):
        return JsonResponse({"detail": "Forbidden"}, status=403)

    contests = (
        Contest.objects.filter(course=course)
        .select_related("course__section", "course__owner", "created_by")
        .annotate(problems_count=Count("problems"))
        .order_by("position", "-created_at")
    )
    is_teacher = _course_is_teacher(course, request.user)
    items = []
    for contest in contests:
        if not contest.is_visible_to(request.user):
            continue
        items.append(
            {
                "id": contest.id,
                "position": contest.position,
                "title": contest.title,
                "description": contest.description,
                "course": contest.course_id,
                "course_title": contest.course.title if contest.course else None,
                "created_by_id": contest.created_by_id,
                "created_by_username": contest.created_by.username if contest.created_by_id else None,
                "is_published": contest.is_published,
                "access_type": contest.access_type,
                "approval_status": contest.approval_status,
                "status": contest.status,
                "is_rated": contest.is_rated,
                "scoring": contest.scoring,
                "registration_type": contest.registration_type,
                "problems_count": contest.problems_count,
                "access_token": contest.access_token
                if contest.access_type == Contest.AccessType.LINK and (is_teacher or is_admin)
                else None,
                **_contest_timing_payload(contest),
            }
        )
    return JsonResponse({"items": items}, status=200)


@login_required
def reorder_course_contests(request, course_id):
    if request.method != "POST":
        return JsonResponse({"detail": "Method not allowed"}, status=405)

    course = get_object_or_404(
        Course.objects.prefetch_related("participants__user"),
        pk=course_id,
    )
    if not _course_is_teacher(course, request.user):
        return JsonResponse({"detail": "Only course teachers can reorder contests"}, status=403)

    try:
        import json

        payload = json.loads(request.body or "{}")
    except Exception:
        return JsonResponse({"detail": "Invalid JSON payload"}, status=400)

    contest_ids = payload.get("contest_ids") or []
    if not isinstance(contest_ids, list) or not contest_ids:
        return JsonResponse({"detail": "contest_ids must be a non-empty list"}, status=400)
    try:
        contest_ids_int = [int(cid) for cid in contest_ids]
    except (TypeError, ValueError):
        return JsonResponse({"detail": "contest_ids must contain integers"}, status=400)

    contests = list(Contest.objects.filter(course=course).order_by("position", "-created_at", "-id"))
    by_id = {c.id: c for c in contests}
    existing_order = [c.id for c in contests]

    requested = [cid for cid in contest_ids_int if cid in by_id]
    if not requested:
        return JsonResponse({"detail": "No provided contest_ids belong to this course"}, status=400)

    remaining = [cid for cid in existing_order if cid not in set(requested)]
    new_order = requested + remaining
    for idx, cid in enumerate(new_order):
        by_id[cid].position = idx
    Contest.objects.bulk_update([by_id[cid] for cid in new_order], ["position"])

    return JsonResponse({"course_id": course.id, "contest_ids": new_order}, status=200)


@login_required
def update_course(request, course_id):
    if request.method != "POST":
        return JsonResponse({"detail": "Method not allowed"}, status=405)

    course = get_object_or_404(Course, pk=course_id)
    is_admin = bool(is_platform_admin(request.user) or request.user.is_staff or request.user.is_superuser)
    if not (is_admin or course.owner_id == request.user.id):
        return JsonResponse({"detail": "Only course owner can update course"}, status=403)

    try:
        import json
        payload = json.loads(request.body or "{}")
    except Exception:
        return JsonResponse({"detail": "Invalid JSON payload"}, status=400)

    update_fields = []
    if "title" in payload:
        course.title = str(payload.get("title") or "").strip()
        update_fields.append("title")
    if "description" in payload:
        course.description = str(payload.get("description") or "")
        update_fields.append("description")
    if "is_open" in payload:
        course.is_open = bool(payload.get("is_open"))
        update_fields.append("is_open")

    if not update_fields:
        return JsonResponse({"detail": "No fields to update"}, status=400)

    course.save(update_fields=update_fields)
    return JsonResponse(
        {
            "id": course.id,
            "title": course.title,
            "description": course.description,
            "is_open": course.is_open,
        },
        status=200,
    )


@login_required
def delete_course(request, course_id):
    if request.method not in {"POST", "DELETE"}:
        return JsonResponse({"detail": "Method not allowed"}, status=405)

    course = get_object_or_404(Course, pk=course_id)
    is_admin = bool(is_platform_admin(request.user) or request.user.is_staff or request.user.is_superuser)
    if not (is_admin or course.owner_id == request.user.id):
        return JsonResponse({"detail": "Only course owner can delete course"}, status=403)

    course.delete()
    return JsonResponse({"success": True, "deleted_id": course_id}, status=200)


@login_required
def update_course_participants(request, course_id):
    if request.method != "POST":
        return JsonResponse({"detail": "Method not allowed"}, status=405)

    course = get_object_or_404(Course.objects.prefetch_related("participants__user"), pk=course_id)
    is_admin = bool(is_platform_admin(request.user) or request.user.is_staff or request.user.is_superuser)
    if not (is_admin or course.owner_id == request.user.id):
        return JsonResponse({"detail": "Only course owner can manage participants"}, status=403)

    try:
        import json

        payload = json.loads(request.body or "{}")
    except Exception:
        return JsonResponse({"detail": "Invalid JSON payload"}, status=400)

    teacher_usernames = payload.get("teacher_usernames") or []
    student_usernames = payload.get("student_usernames") or []
    allow_role_update = bool(payload.get("allow_role_update", True))

    if not isinstance(teacher_usernames, list) or not isinstance(student_usernames, list):
        return JsonResponse(
            {"detail": "teacher_usernames and student_usernames must be lists"},
            status=400,
        )

    teacher_usernames = [str(u).strip() for u in teacher_usernames if str(u).strip()]
    student_usernames = [str(u).strip() for u in student_usernames if str(u).strip()]

    users = list(User.objects.filter(username__in=set(teacher_usernames + student_usernames)))
    user_by_username = {u.username: u for u in users}

    missing = sorted(
        {u for u in set(teacher_usernames + student_usernames) if u not in user_by_username}
    )
    if missing:
        return JsonResponse({"detail": "Some users not found", "missing": missing}, status=400)

    from ..services.course_service import add_users_to_course

    result = add_users_to_course(
        course=course,
        teachers=[user_by_username[u] for u in teacher_usernames],
        students=[user_by_username[u] for u in student_usernames],
        allow_role_update=allow_role_update,
    )

    def _summarize(participants):
        return [
            {
                "user_id": p.user_id,
                "username": p.user.username,
                "role": p.role,
                "is_owner": p.is_owner,
            }
            for p in participants
        ]

    return JsonResponse(
        {"course_id": course.id, "created": _summarize(result["created"]), "updated": _summarize(result["updated"])},
        status=200,
    )


@login_required
def remove_course_participants(request, course_id):
    if request.method != "POST":
        return JsonResponse({"detail": "Method not allowed"}, status=405)

    course = get_object_or_404(Course, pk=course_id)
    is_admin = bool(is_platform_admin(request.user) or request.user.is_staff or request.user.is_superuser)
    if not (is_admin or course.owner_id == request.user.id):
        return JsonResponse({"detail": "Only course owner can manage participants"}, status=403)

    try:
        import json

        payload = json.loads(request.body or "{}")
    except Exception:
        return JsonResponse({"detail": "Invalid JSON payload"}, status=400)

    usernames = payload.get("usernames") or []
    if not isinstance(usernames, list) or not usernames:
        return JsonResponse({"detail": "usernames must be a non-empty list"}, status=400)
    usernames = [str(u).strip() for u in usernames if str(u).strip()]
    users = list(User.objects.filter(username__in=set(usernames)))
    by_name = {u.username: u for u in users}
    missing = sorted({u for u in set(usernames) if u not in by_name})
    if missing:
        return JsonResponse({"detail": "Some users not found", "missing": missing}, status=400)

    if course.owner_id and course.owner.username in set(usernames):
        return JsonResponse({"detail": "Cannot remove course owner"}, status=400)

    deleted, _ = CourseParticipant.objects.filter(course=course, user__in=users).delete()
    return JsonResponse({"course_id": course.id, "removed": usernames, "deleted": deleted}, status=200)
